File: term_1 (C++, Algos and Data Structs)/homework_6/6.1/stack.cpp
#include "stack.h"
#include <iostream>

struct Element
{
	FlagType value;
	Element *next;
};

struct Stack
{
	Element *head;
};

Stack *createStack()
{
	return new Stack{ nullptr };
}

void push(Stack *&stack, FlagType value)
{
	Element *newElement = new Element{ value, stack->head };
	stack->head = newElement;
}

bool isEmpty(Stack *&stack)
{
	if (!stack->head)
	{
		return true;
	}
	return false;
}

FlagType pop(Stack *&stack, bool &isEmptyStack)
{
	if (isEmpty(stack))
	{
		isEmptyStack = true;
		return -1;
	}
	Element *position = stack->head;
	FlagType popElement = position->value;
	stack->head = position->next;
	delete position;
	return popElement;
}

void deleteStack(Stack *&stack)
{
	bool isEmptyStack = false;
	while (!isEmptyStack)
	{
		pop(stack, isEmptyStack);
	}
	delete stack;
	stack = nullptr;
}

void printStack(Stack *&stack)
{
	Element *position = stack->head;
	while (position)
	{
		std::cout << position->value << std::endl;
		position = position->next;
	}
}

File: term_2 (C#, OOP)/homework_7/7.2(Set)/7.2(Set)/Stack.cs
namespace StackNamespace
{
    /// <summary>
    /// Стек на списках.
    /// </summary>
    public class Stack<T>
    {
        /// <summary>
        /// Элемент стека.
        /// </summary>
        private class StackElement
        {
            /// <summary>
            /// Значение элемента стека.
            /// </summary>
            public T Value { get; set; }

            /// <summary>
            /// Следующий за данным элемент стека.
            /// </summary>
            public StackElement Next { get; set; }

            /// <summary>
            /// Конструктор экземпляра класса <see cref="StackElement"/>.
            /// </summary>
            /// <param name="value"></param>
            /// <param name="next"></param>
            public StackElement(T value, StackElement next)
            {
                this.Next = next;
                this.Value = value;
            }
        }

        /// <summary>
        /// Головной элемент стека.
        /// </summary>
        private StackElement head;

        /// <summary>
        /// Конструктор экземпляра класса <see cref="Stack"/>.
        /// </summary>
        public Stack()
        {
        }

        /// <summary>
        /// Добавление элемента в стек.
        /// </summary>
        /// <param name="value">Значение добавляемого элемента.</param>
        public void Push(T value)
        {
            StackElement newElement = new StackElement(value, head);
            head = newElement;
        }

        /// <summary>
        /// Иъятие головного элемента стека.
        /// </summary>
        /// <returns>Значение головного элемента.</returns>
        public T Pop()
        {
            if (IsEmpty())
            {
                throw new EmptyStackExeption("попытка обращения к элементу путого стека");
            }

            T popElement = head.Value;
            head = head.Next;
            return popElement;
        }

        /// <summary>
        /// Чтение головного элемента стека. 
        /// </summary>
        /// <returns>Значение головного элемента.</returns>
        public T Peek()
        {
            if (IsEmpty())
            {
                throw new EmptyStackExeption("попытка обращения к элементу путого стека");
            }

            return head.Value;
        }

        /// <summary>
        /// Проверка стека на пустоту.
        /// </summary>
        /// <returns>True если пуст.</returns>
        public bool IsEmpty() => head == null;
    }
}
